mcs/class/System/System.Net.Sockets/Socket.cs

```diff
@@ -968,13 +985,26 @@ public IAsyncResult BeginAccept(AsyncCallback callback, object state)
 			if (!is_bound || !is_listening)
 				throw new InvalidOperationException ();
 
-			SocketAsyncResult sockares = new SocketAsyncResult (this, state, callback, SocketOperation.Accept);
+			SocketAsyncResult sockares = new SocketAsyncResult (this, callback, state, SocketOperation.Accept);
 
-			QueueSocketAsyncResult (readQ, sockares.Worker, sockares);
+			QueueIOSelectorJob (readQ, sockares.handle, new IOSelectorJob (IOOperation.Read, s => BeginAcceptCallback ((SocketAsyncResult) s), sockares));
 
 			return sockares;
 		}
 
+		static void BeginAcceptCallback (SocketAsyncResult sockares)
+		{
+			Socket socket = null;
+			try {
+				socket = sockares.socket.Accept ();
+			} catch (Exception e) {
```

Comment:
Because it happens at different times: AcceptAsyncCallback is called after BeginAcceptCallback. The AcceptAsync pattern, is using the same mechanism as the Begin/EndAccept mechanism, except that we control more of the work, so we can optimize it further (less allocation, ...). You can observe that in the way QueueIOSelectorJob is called in AcceptAsync:
```
QueueIOSelectorJob (readQ, e.socket_async_result.handle, new IOSelectorJob (IOOperation.Read, s => BeginAcceptCallback ((SocketAsyncResult) s), e.socket_async_result));
```
The callback to call when there is something on the socket is `BeginAcceptCallback` (exactly the same one as in `BeginAccept`), but the difference lies in the callback called when `BeginAcceptCallback` has finished: it calls `AcceptAsyncCallback`, instead of a user defined one, and `AcceptAsyncCallback` is going to call `EndAccept`.